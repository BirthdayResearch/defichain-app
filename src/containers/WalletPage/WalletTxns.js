import React, { Component } from 'react'
import {
  Card,
  Table
} from 'reactstrap';
import {
  MdArrowUpward,
  MdArrowDownward
} from "react-icons/md";
import styles from './WalletTxns.module.scss';

function TxnTypeIcon(type) {
  let Icon;
  if (type === 'Received') {
    Icon = MdArrowDownward;
  } else {
    Icon = MdArrowUpward;
  }

  return <Icon className={styles.typeIcon} />;
}

class WalletTxns extends Component {
  state = {
    txns: [
      { id: 0, type: 'Received', time: 'Feb 19, 2:03 pm', hash: 'c9a59be5d9f453229f519ab3c5289c', value: 100, unit: 'DFI' },
      { id: 1, type: 'Received', time: 'Feb 19, 2:03 pm', hash: 'c9a59be5d9f453229f519ab3c5289c', value: 100, unit: 'DFI' },
      { id: 2, type: 'Received', time: 'Feb 19, 2:03 pm', hash: 'c9a59be5d9f453229f519ab3c5289c', value: 100, unit: 'DFI' },
      { id: 3, type: 'Sent', time: 'Feb 19, 2:03 pm', hash: 'c9a59be5d9f453229f519ab3c5289c', value: 100, unit: 'DFI' },
      { id: 4, type: 'Received', time: 'Feb 19, 2:03 pm', hash: 'c9a59be5d9f453229f519ab3c5289c', value: 100, unit: 'DFI' },
      { id: 5, type: 'Received', time: 'Feb 19, 2:03 pm', hash: 'c9a59be5d9f453229f519ab3c5289c', value: 100, unit: 'DFI' },
      { id: 6, type: 'Sent', time: 'Feb 19, 2:03 pm', hash: 'c9a59be5d9f453229f519ab3c5289c', value: 100, unit: 'DFI' },
      { id: 7, type: 'Received', time: 'Feb 19, 2:03 pm', hash: 'c9a59be5d9f453229f519ab3c5289c', value: 100, unit: 'DFI' }
    ]
  };

  render() {
    return (
      <Card className="table-responsive-md">
        <Table className={styles.table}>
          <tbody>
            {this.state.txns.map((txn) => (
              <tr key={txn.id}>
                <td className={styles.typeIcon}>
                  {TxnTypeIcon(txn.type)}
                </td>
                <td>
                  <div className={styles.type}>
                    {txn.type}
                  </div>
                  <div className={styles.time}>
                    {txn.time}
                  </div>
                </td>
                <td>
                  <div className={styles.hash}>
                    {txn.hash}
                  </div>
                </td>
                <td>
                  <div className={styles.value}>
                    {txn.value} <span className={styles.unit}>{txn.unit}</span>
                  </div>
                </td>
              </tr>
            ))}
          </tbody>
        </Table>
      </Card>
    )
  }
}

export default WalletTxns;