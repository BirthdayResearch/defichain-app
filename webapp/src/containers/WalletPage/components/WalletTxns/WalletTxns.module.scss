// App and Bootstrap variables required
@import "~bootstrap/scss/functions";
@import "~bootstrap/scss/variables";
@import "~bootstrap/scss/mixins";
@import "../../../../scss/variables";
@import "../../../../scss/bootstrap/variables";

.table {
  margin-bottom: 0;

  td {
    vertical-align: middle;
  }

  tr:first-child td,
  tr:first-child th {
    border-top: none;
  }
}

.typeIcon {
  width: 24px;
  height: 24px;
  color: $secondary;
}

.type,
.time {
  white-space: nowrap;
}

.hash {
  font-size: 0.875rem;
  letter-spacing: 0.25px;
}

.amount {
  text-align: right;
  letter-spacing: 0.44px;
}

.unit {
  color: $secondary;
}
