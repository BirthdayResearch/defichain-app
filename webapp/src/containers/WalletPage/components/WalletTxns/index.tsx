import React, { useState, useEffect } from 'react';
import { Card, Table, CardBody } from 'reactstrap';
import { connect } from 'react-redux';
import { MdArrowUpward, MdArrowDownward } from 'react-icons/md';
import styles from './WalletTxns.module.scss';
import { I18n } from 'react-redux-i18n';
import { fetchWalletTxnsRequest } from '../../reducer';
import { WALLET_TXN_PAGE_SIZE } from '../../../../constants';
import Pagination from '../../../../components/Pagination';
import { getAmountInSelectedUnit } from '../../../../utils/utility';

interface WalletTxnsProps {
  unit: string;
  walletTxns: {
    txnId: string;
    category: string;
    time: string;
    amount: number;
    unit: string;
  }[];
  walletTxnCount: number;
  fetchWalletTxns: (currentPage: number, pageSize: number) => void;
}

interface WalletTxnsState {
  currentPage: number;
  pageSize: number;
}

const WalletTxns: React.FunctionComponent<WalletTxnsProps> = (
  props: WalletTxnsProps
) => {
  const [currentPage, handlePageChange] = useState(1);
  const pageSize = WALLET_TXN_PAGE_SIZE;

  useEffect(() => {
    props.fetchWalletTxns(currentPage, pageSize);
  }, []);

  const getTxnsTypeIcon = (type: string) => {
    if (type === 'send') {
      return <MdArrowUpward className={styles.typeIcon} />;
    }
    return <MdArrowDownward className={styles.typeIcon} />;
  };

  const fetchData = (index: number) => {
    props.fetchWalletTxns(index, pageSize);
    handlePageChange(index);
  };

  const { walletTxnCount: total, walletTxns } = props;
  const pagesCount = Math.ceil(total / pageSize);
  const from = (currentPage - 1) * pageSize;
  const to = Math.min(total, currentPage * pageSize);

  return (
    <section className='mb-5'>
      <h2>{I18n.t('containers.wallet.walletPage.transactions')}</h2>
      {pagesCount ? (
        <>
          <Card className={`${styles.card} table-responsive-md`}>
            <Table className={styles.table}>
              <thead>
                <tr>
                  <th></th>
                  <th>{I18n.t('containers.wallet.walletTxns.time')}</th>
                  <th className={styles.amount}>
                    {I18n.t('containers.wallet.walletTxns.amount')}
                  </th>
                  <th>{I18n.t('containers.wallet.walletTxns.hash')}</th>
                </tr>
              </thead>
              <tbody>
                {walletTxns.map((txn, index) => (
                  <tr key={`${txn.txnId}-${index}`}>
                    <td className={styles.typeIcon}>
                      {getTxnsTypeIcon(txn.category)}
                    </td>
                    <td>
                      <div className={styles.time}>{txn.time}</div>
                    </td>
                    <td>
                      <div className={styles.amount}>
                        {getAmountInSelectedUnit(
                          txn.amount,
                          props.unit,
                          txn.unit
                        )}{' '}
                        <span className={styles.unit}>{props.unit}</span>
                      </div>
                    </td>
                    <td>
                      <div className={styles.hash}>{txn.txnId}</div>
                    </td>
                  </tr>
                ))}
              </tbody>
            </Table>
          </Card>
          <Pagination
            label={I18n.t('containers.wallet.walletPage.paginationRange', {
              to,
              total,
              from: from + 1,
            })}
            currentPage={currentPage}
            pagesCount={pagesCount}
            handlePageClick={fetchData}
          />
        </>
      ) : (
        <Card className='table-responsive-md'>
          <CardBody>
            {I18n.t('containers.wallet.walletTxns.noTransactions')}
          </CardBody>
        </Card>
      )}
    </section>
  );
};

const mapStateToProps = state => {
  const { settings, wallet } = state;
  return {
    unit: settings.appConfig.unit,
    walletTxns: wallet.walletTxns,
    walletTxnCount: wallet.walletTxnCount,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    fetchWalletTxns: (currentPage, pageSize) =>
      dispatch(fetchWalletTxnsRequest({ currentPage, pageSize })),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(WalletTxns);
