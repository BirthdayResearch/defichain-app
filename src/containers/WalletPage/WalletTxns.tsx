import React, { Component } from "react";
import { Card, Table } from "reactstrap";
import { MdArrowUpward, MdArrowDownward } from "react-icons/md";
import styles from "./WalletTxns.module.scss";
import { WalletTxnsProps, WalletTxnsState } from "./WalletPage.interface";

class WalletTxns extends Component<WalletTxnsProps, WalletTxnsState> {
  state = {
    txns: [
      {
        id: 0,
        type: "Received",
        time: "Feb 19, 2:03 pm",
        hash: "c9a59be5d9f453229f519ab3c5289c",
        amount: 100,
        unit: "DFI"
      },
      {
        id: 1,
        type: "Received",
        time: "Feb 19, 2:03 pm",
        hash: "c9a59be5d9f453229f519ab3c5289c",
        amount: 100,
        unit: "DFI"
      },
      {
        id: 2,
        type: "Received",
        time: "Feb 19, 2:03 pm",
        hash: "c9a59be5d9f453229f519ab3c5289c",
        amount: 100,
        unit: "DFI"
      },
      {
        id: 3,
        type: "Sent",
        time: "Feb 19, 2:03 pm",
        hash: "c9a59be5d9f453229f519ab3c5289c",
        amount: 100,
        unit: "DFI"
      },
      {
        id: 4,
        type: "Received",
        time: "Feb 19, 2:03 pm",
        hash: "c9a59be5d9f453229f519ab3c5289c",
        amount: 100,
        unit: "DFI"
      },
      {
        id: 5,
        type: "Received",
        time: "Feb 19, 2:03 pm",
        hash: "c9a59be5d9f453229f519ab3c5289c",
        amount: 100,
        unit: "DFI"
      },
      {
        id: 6,
        type: "Sent",
        time: "Feb 19, 2:03 pm",
        hash: "c9a59be5d9f453229f519ab3c5289c",
        amount: 100,
        unit: "DFI"
      },
      {
        id: 7,
        type: "Received",
        time: "Feb 19, 2:03 pm",
        hash: "c9a59be5d9f453229f519ab3c5289c",
        amount: 100,
        unit: "DFI"
      }
    ]
  };

  TxnTypeIcon = type => {
    let Icon;
    if (type === "Received") {
      Icon = MdArrowDownward;
    } else {
      Icon = MdArrowUpward;
    }

    return <Icon className={styles.typeIcon} />;
  };

  render() {
    return (
      <Card className="table-responsive-md">
        <Table className={styles.table}>
          <thead>
            <tr>
              <th></th>
              <th>Time</th>
              <th className={styles.amount}>Amount</th>
              <th>Hash</th>
            </tr>
          </thead>
          <tbody>
            {this.state.txns.map(txn => (
              <tr key={txn.id}>
                <td className={styles.typeIcon}>
                  {this.TxnTypeIcon(txn.type)}
                </td>
                <td>
                  <div className={styles.time}>{txn.time}</div>
                </td>
                <td>
                  <div className={styles.amount}>
                    {txn.amount} <span className={styles.unit}>{txn.unit}</span>
                  </div>
                </td>
                <td>
                  <div className={styles.hash}>{txn.hash}</div>
                </td>
              </tr>
            ))}
          </tbody>
        </Table>
      </Card>
    );
  }
}

export default WalletTxns;
