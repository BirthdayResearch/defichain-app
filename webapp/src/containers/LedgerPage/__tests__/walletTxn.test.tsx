import React from 'react';
import WalletTxns from '../components/WalletTxns';
import { Provider } from 'react-redux';
import store from '../../../app/rootStore';
import { mount } from 'enzyme';

describe('WalletTxns component', () => {
  it('should check for snapshot', () => {
    const tokenSymbol = 'DFI';
    const wrapper = mount(
      <Provider store={store}>
        <WalletTxns tokenSymbol={tokenSymbol} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
