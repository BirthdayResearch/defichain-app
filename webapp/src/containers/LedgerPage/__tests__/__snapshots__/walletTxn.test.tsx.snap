// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletTxns component should check for snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(WalletTxns)>
    <WalletTxns
      fetchWalletTxns={[Function]}
      stopPagination={true}
      unit="DFI"
      walletTxnCount={0}
      walletTxns={Array []}
    >
      <section
        className="mb-5"
      >
        <h2>
          Transactions
        </h2>
        <Card
          className="table-responsive-md"
          tag="div"
        >
          <div
            className="table-responsive-md card"
          >
            <CardBody
              tag="div"
            >
              <div
                className="card-body"
              >
                No transactions available
              </div>
            </CardBody>
          </div>
        </Card>
      </section>
    </WalletTxns>
  </Connect(WalletTxns)>
</Provider>
`;
